mh/pat.py

```diff
@@ -2412,15 +2439,31 @@ def handle(self):
         self.server.info("Handle client")
         self.server.add_to_debug(self)
         self.session = Session(self)
-
+        exception_raised = False
+        
+        # Temporary workaround for Dolphin bug
+        time.sleep(2)
         self.sendReqConnection()
         try:
             self.handle_client()
-            self.session.disconnect()
         except Exception as e:
+            exception_raised = True
             self.server.error(traceback.format_exc())
             self.send_error("{}: {}".format(type(e).__name__, str(e)))
-            self.session.delete()
+        finally:
+            self.session.disconnect()
+            if not self.finished or exception_raised:
+                try:
+                    self.notify_layer_departure()
+                except Exception:
+                    # We tried to comunicate the error to the client but it
+                    # seems that the exception was caused by a connection
+                    # error and in that case there is nothing else we can do
+                    # to communicate that error to the client
+                    pass
```

Comment:
We need the `exception_raise` var, because if a non socket related exception is catch, `self.finished` would be `False` but when the exception message is sent to the client through the `sendNtcShut` packet (which is called through `self.send_error`) the `self.finish` method would also be called which would set `self.finished = True` which would make this:
```py
if not self.finished:
(...)
```

not execute because `self.finished == False`, but it's a valid path because we need to correctly remove the player from the layers. 

I do agree about not needing the `finally` block and about calling `session.layer_end` inside the `session.delete` method

Reply:
@kbarkevich You missed this comment